pennylane/operation.py

```diff
@@ -1887,3 +1863,23 @@ def is_trainable(obj):
     """Returns ``True`` if any of the parameters of an operator is trainable
     according to ``qml.math.requires_grad``."""
     return any(qml.math.requires_grad(p) for p in obj.parameters)
+
+
+@qml.BooleanFn
+def defines_diagonalizing_gates(obj):
+    """Returns ``True`` if an operator defines the diagonalizing
+    gates are defined."""
+
+    # We need to stop the tape so that the diagonalizing gates
+    # are not queued when we check if they are defined.
+    # Reassess this when queueing system is overhauled
+    tape = qml.tape.get_active_tape()
+    if tape is not None:
+        with tape.stop_recording():
+            dgates = obj.diagonalizing_gates()
+    else:
+        dgates = obj.diagonalizing_gates()
+
+    if dgates is None:
+        return False
+    return True
```

Comment:
```suggestion
    with qml.tape.stop_recording():
        dgates = obj.diagonalizing_gates()

    return dgates is not None
```

`qml.tape.stop_recording` fixes a lot of the weird queuing issues we previously had :)

Alternatively, it can be used as a decorator:

```python
dgates = stop_recording()(obj.diagonalizing_gates)()
```